pkg/progress/progress.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package progress
+
+import (
+	"sync"
+	"time"
+)
+
+// Manager is used to maintain the progresses we are care about.
+type Manager struct {
+	sync.RWMutex
+	progesses map[string]*progressIndicator
+}
+
+// NewManager creates a new Manager.
+func NewManager() *Manager {
+	return &Manager{
+		progesses: make(map[string]*progressIndicator),
+	}
+}
+
+// progressIndicator reflects a specified progress.
+type progressIndicator struct {
+	total          float64
+	current        float64
+	startTime      time.Time
+	lastUpdateTime time.Time
+	speedPerSec    float64
+}
+
+// AddProgressIndicator adds a progress into manager.
+func (m *Manager) AddProgressIndicator(progress string, total float64) {
+	m.Lock()
+	defer m.Unlock()
+	m.progesses[progress] = &progressIndicator{
```

Comment:
Only two kinds of situations the function will be called: 1. The API triggers. In this case, it will prevent calling twice in the same store. 2. In the start process, we only call it once.

Reply:
How about merge the function `AddProgressIndicator`, `Update` into 1
```
// AddProgressIndicator inits or updates the progress of a given name. If the progress with the given name not exist, update it,
// returns true when the progress already exist.
func (m *Manager) AddProgressIndicator(progress string, current float64) (exist bool) {
	m.Lock()
	defer m.Unlock()
	curPro, ok := m.progesses[progress]
	if !ok {
		m.progesses[progress] = &progressIndicator{
			total:     current,
			startTime: time.Now(),
		}
		return ok
	}
	curPro.current = current
	if curPro.total < current {
		curPro.total = current
	}
	m.progesses[progress].speedPerSec = (m.progesses[progress].total - m.progesses[progress].current) / time.Since(m.progesses[progress].startTime).Seconds()
	return ok
}
```